legion/engine/core/particles/particlesystem.cpp

```diff
@@ -0,0 +1,117 @@
+#include <core/particles/particles.hpp>
+#include <core/common/debugrendering.hpp>
+
+
+namespace legion::core
+{
+    void ParticleSystem::setup()
+    {
+        log::filter(log::severity_debug);
+        log::debug("ParticleSystem setup");
+
+        bindToEvent<events::component_creation<particle_emitter>, &ParticleSystem::emitter_setup>();
+    }
+
+    void ParticleSystem::shutdown()
+    {
+        log::debug("ParticleSystem shutdown");
+    }
+
+    void ParticleSystem::update(time::span deltaTime)
+    {
+        ecs::filter<particle_emitter> filter;
+        for (auto& ent : filter)
+        {
+            auto& emitter = ent.get_component<particle_emitter>().get();
+            if (emitter.currentParticleCount < emitter.maxSpawnCount)
+            {
+                emitter.elapsedTime += deltaTime;
+                if (emitter.elapsedTime >= emitter.spawnInterval)
+                {
+                    emit(emitter, emitter.spawnRate);
+                    emitter.elapsedTime = 0;
+                }
+            }
+            maintanence(emitter, deltaTime);
+        }
+    }
+
+    void ParticleSystem::emitter_setup(L_MAYBEUNUSED events::component_creation<particle_emitter>& event)
+    {
+        ecs::filter<particle_emitter> filter;
+        for (auto& ent : filter)
+        {
+            auto emitter = ent.get_component<particle_emitter>();
+            for (size_type i = 0; i < emitter->maxSpawnCount; i++)
+                emitter->set_alive(i, false);
```

Comment:
you set all particles of all emitters to dead, even if the particles don't exist?